include/nbl/builtin/hlsl/type_traits.hlsl

```diff
@@ -4,13 +4,425 @@
+    is_same<T, bool>::value ||
+    is_same<T, uint16_t>::value ||
+    is_same<T, uint32_t>::value ||
+    is_same<T, uint64_t>::value ||
+    is_same<T, int16_t>::value ||
+    is_same<T, int32_t>::value ||
+    is_same<T, int64_t>::value
+> {};
+
+template<class T>
+struct is_floating_point : bool_constant<
+    is_same<T, half>::value ||
+    is_same<T, float>::value ||
+    is_same<T, double>::value
+> {};
+
+template<class T>
+struct is_signed : bool_constant<
+    is_same<T, int16_t>::value ||
+    is_same<T, int32_t>::value ||
+    is_same<T, int64_t>::value ||
+    is_floating_point<T>::value
+> {};
+
+}
+
+template<class T> 
+struct is_unsigned : impl::base_type_forwarder<impl::is_unsigned, T> {};
+
+template<class T> 
+struct is_integral : impl::base_type_forwarder<impl::is_integral, T> {};
+
+template<class T> 
+struct is_floating_point : impl::base_type_forwarder<impl::is_floating_point, T> {};
+
+template<class T> 
+struct is_signed : impl::base_type_forwarder<impl::is_signed, T> {};
+
+template<class T>
+struct is_scalar : bool_constant<
+    impl::is_integral<T>::value || 
+    impl::is_floating_point<T>::value
+> {};
+
+template<class T>
+struct is_const : bool_constant<false> {};
+
+template<class T>
+struct is_const<const T> : bool_constant<true> {};
+
+template<class T>
+struct is_volatile : bool_constant<false> {};
+
+template<class T>
+struct is_volatile<volatile T> : bool_constant<true> {};
+
+template<class>
+struct is_trivial : bool_constant<true> {};
+
+template<class>
+struct is_trivially_copyable : bool_constant<true> {};
+
+// this implementation is fragile
+template<class T>
+struct is_empty : bool_constant<0==sizeof(T)> {};
+
+template<class>
+struct is_polymorphic : bool_constant<false> {};
+
+template<class>
+struct is_abstract : bool_constant<false> {};
+
+template<class>
+struct is_final : bool_constant<false> {};
+
+template <class T>
+struct is_fundamental : bool_constant<
+    is_scalar<T>::value || 
+    is_void<T>::value
+> {};
+
+template <class T>
+struct is_compound : bool_constant<!is_fundamental<T>::value> {};
+
+template <class T>
+struct is_aggregate : is_compound<T> {};
+
+
+// need this crutch because we can't make `#define typeid` work both on expression and types
+template<typename T>
+struct typeid_t;
+
+namespace impl
+{
+template<uint32_t N>
+struct encoder
+{
+    uint32_t arr[N];
+};
+template<uint32_t encoded_typeid>
+struct decltype_t;
 }
 
+
+// DXC doesn't support linking SPIR-V so this will always work I guess?
+// split because we won't be able to use `typeid` or `decltype` on functions until https://github.com/microsoft/hlsl-specs/issues/100
+#define NBL_REGISTER_TYPEID(T) template<> struct typeid_t<T> : integral_constant<uint32_t,__COUNTER__> {}
+
+#define NBL_REGISTER_OBJ_TYPE(T) NBL_REGISTER_TYPEID(T); \
+namespace impl \
+{ \
+encoder<typeid_t<T>::value> encode_typeid(T); \
+template<> struct decltype_t<sizeof(encoder<typeid_t<T>::value>)/4> \
+{ \
+  using type = T; \
+}; \
+} 
```

Comment:
these need to be fully qualified instead of in `namespace` so I guess its `template struct ::nbl::hlsl::type_traits::type_t : ::nbl::hlsl::type_traits::integral_constant...`, etc.